Plugin.pm

```diff
@@ -617,29 +620,70 @@ sub QobuzArtist {
 		# use album list if it was returned in the artist lookup
 		if ($artist->{albums}) {
 			my $albums = [];
+			my $numAlbums = 0;
+			my $numEps = 0;
+			my $numSingles = 0;
+			
+			# group by release type if requested
+			if ($groupByReleaseType) {
+				for my $album ( @{$artist->{albums}->{items}} ) {
+					next if $args->{artistId} && $album->{artist}->{id} != $args->{artistId};					
+					if ($album->{duration} >= 1800 || $album->{tracks_count} > 6) {
+						$album->{release_type} = "Album";
+						$numAlbums++;
+					} elsif ($album->{tracks_count} < 4) {
+						$album->{release_type} = "Single";
+						$numSingles++;
+					} else {
+						$album->{release_type} = "Ep";
+						$numEps++;
+					}
+				}
+			}
 
 			# sort by release date if requested
 			my $sortByDate = $prefs->get('sortArtistAlbums');
 
 			$artist->{albums}->{items} = [ sort {
 				if ($sortByDate) {
-					return $sortByDate == 1 ? $b->{released_at}*1 <=> $a->{released_at}*1 : $a->{released_at}*1 <=> $b->{released_at}*1;
+					return $sortByDate == 1 ? ( $a->{release_type} cmp $b->{release_type} || $b->{released_at}*1 <=> $a->{released_at}*1 )
+											: ( $a->{release_type} cmp $b->{release_type} || $a->{released_at}*1 <=> $b->{released_at}*1 );
 				}
 				else {
-					# push singles and EPs down the list
-					if ( ($a->{tracks_count} >= 4 && $b->{tracks_count} < 4) || ($a->{tracks_count} < 4 && $b->{tracks_count} >=4) ) {
-						return $b->{tracks_count} <=> $a->{tracks_count};
-					}
-
-					return lc($a->{title}) cmp lc($b->{title});
+					return $a->{release_type} cmp $b->{release_type} || lc($a->{title}) cmp lc($b->{title});
 				}
 
 			} @{$artist->{albums}->{items} || []} ];
 
+			my $lastReleaseType = "";
+
 			for my $album ( @{$artist->{albums}->{items}} ) {
 				next if $args->{artistId} && $album->{artist}->{id} != $args->{artistId};
+				if ($album->{release_type} ne $lastReleaseType) {
+					$lastReleaseType = $album->{release_type};
+					my $relType = "";
+					my $relNum = 0;
+					if ($lastReleaseType eq "Album") {
```

Comment:
I'd do case insensitive comparisons. It's too easy to get it wrong.

Reply:
Are you aware that these literals are temporary values assigned to the album release types while traversing the array a few lines previous to this in preparation for the sort? I could have used 'a', 'b' and 'c' or even 1, 2 and 3 but I thought the use of the literals would be more self-documenting. I'm open to specific suggestions though.

Reply:
Understood! But as it's easy to get a literal string comparison wrong without noticing I suggested using constants: those would throw an error upon startup if you introduced a typo.